File: nvpl_blas/f77/drotm.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DROTM 
!
!******************************************************************************
      program   DROTM_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic        abs

      integer          n, incx, incy, i
      parameter        (n=5, incx=1, incy=1)
      integer          xsize, ysize
      parameter        (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      double precision x(xsize), y(ysize), P(5)
      data             (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/,
     &                 (y(i), i=1,ysize)/ysize*2.0/,
     &                 (P(i), i=1,5)/1.0,0.25,-1.0,1.0,0.5/

*       External Subroutines
      external        print_dvector 
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, incy
      call print_dvector(x,n,incx,'X')
      call print_dvector(y,n,incy,'Y')
      call print_dvector(P,5,1,'P')
*
*      Call DROTM subroutine
      call DROTM(n,x,incx,y,incy,P)
*
      print*
      print 101
      call print_dvector(x,n,incx,'X')
      call print_dvector(y,n,incy,'Y')

      stop
 99   format('Example: DROTM for applying Givens rotation matrix')
 100  format('#### args: n=',i1,', incx=',i1,', incy=',i1)
 101  format('The result of DROTM: ')
      end

File: nvpl_blas/c/drotm.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_drotm
 *
 ******************************************************************************/
#include "example_helper.h"
#include <math.h> // sqrt

int main() {
    nvpl_int_t N = 5;
    double * X;
    double * Y;
    double P[5] = {1.0f, 0.25f, -1.0f, 1.0f, 0.5f};
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    printf("\nExample: cblas_drotm for applying Givens rotation matrix\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", incy=%" PRId64 "\n", (int64_t)N,
            (int64_t)incX, (int64_t)incY);
    print_dvector(P, 5, 1, "P");

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);

    // allocate memory
    X = (double *)malloc(len_x * sizeof(double));
    Y = (double *)malloc(len_y * sizeof(double));

    // fill data
    fill_dvector(X, N, incX);
    fill_dvector(Y, N, incY);

    // print input data
    print_dvector(X, N, incX, "X");
    print_dvector(Y, N, incY, "Y");

    // call cblas_drotm
    cblas_drotm(N, X, incX, Y, incY, P);

    // print result
    printf("\nThe result of cblas_drotm:\n");
    print_dvector(X, N, incX, "X");
    print_dvector(Y, N, incY, "Y");

    // release memory
    free(X);
    free(Y);
    return 0;
}
